avro-java8/src/main/java/com/fasterxml/jackson/dataformat/avro/java8/deser/BaseTimeJsonDeserializer.java

```diff
@@ -0,0 +1,43 @@
+package com.fasterxml.jackson.dataformat.avro.java8.deser;
+
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JsonDeserializer;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.temporal.ChronoUnit;
+import java.util.concurrent.TimeUnit;
+
+public abstract class BaseTimeJsonDeserializer<T> extends JsonDeserializer<T> {
+  private final TimeUnit resolution;
+  final ZoneId zoneId = ZoneId.of("UTC");
+
+  BaseTimeJsonDeserializer(TimeUnit resolution) {
+    this.resolution = resolution;
+  }
+
+  abstract T fromInstant(Instant input);
+
+  @Override
+  public T deserialize(JsonParser p, DeserializationContext ctxt) throws IOException {
+    final long input = p.getLongValue();
```

Comment:
Should probably add checking for other input types, give more meaningful error message in case underlying physical token is not a 64-bit long (schema mismatch)? And unit test for the same would be useful as well.